statsmodels/regression/tests/test_mixed_glm.py

```diff
@@ -0,0 +1,350 @@
+    test_links = []
+
+    def test_links_derivs(self):
+        for lnk in self.test_links:
+            try:
+                self.like_hess_grad(self.family, lnk)
+            except Exception as e:
+                e.args += ("Link type: " + str(lnk),)
+                raise
+
+    def like_hess_grad(self, fam, lnk):
+        """
+        Test the function, gradient and Hessian of the joint
+        log-likelihood of the random effects and the data.
+        """
+
+        fa = fam(link=lnk)
+
+        np.random.seed(423)
+        endog, exog, groups = gen_mixed(fa.__class__.__name__, 10)
+
+        model = MixedGLM(endog, exog, groups=groups, family=fa)
+
+        # Set some parameters
+        fe_params = np.random.normal(size=2)
+        cov_re = np.abs(np.random.normal(size=(1,1)))
+        scale = 1.0
+
+        self.like_hess_grad_helper(model, fe_params, cov_re, scale)
+
+    def test_links_derivs2(self):
+        for lnk in self.test_links:
+            try:
+                self.like_hess_grad2(self.family, lnk)
+            except Exception as e:
+                e.args += ("Link type: " + str(lnk),)
+                raise
+
+    def like_hess_grad2(self, fam, lnk):
+        """
+        Test the function, gradient and Hessian of the joint
+        log-likelihood of the random effects and the data.
+        """
+
+        fa = fam(link=lnk)
+
+        np.random.seed(423)
+        endog, exog, groups, rs = gen_mixed2(fa.__class__.__name__, 10)
+        exog_re = np.vstack((np.ones(len(groups)),rs)).T
+
+        model = MixedGLM(endog, exog, groups=groups, exog_re=exog_re, family=fa)
+
+        # Set some parameters
+        fe_params = np.random.normal(size=2)
+        cov_re = np.abs(np.random.normal(size=(2,2)))
+        cov_re = np.dot(cov_re.T,cov_re)
+        scale = 1.0
+
+        self.like_hess_grad_helper(model, fe_params, cov_re, scale)
+
+
+    def like_hess_grad_helper(self, model, fe_params, cov_re, scale):
+
+        # Get the log-likelihood function, and its gradient and Hessian.
+        fungradhess = model._gen_joint_like_grad_hess(fe_params, cov_re, scale)
+        fun1 = lambda x : fungradhess(x)[0]
+        grad = lambda x : fungradhess(x)[1]
+        hess = lambda x : fungradhess(x)[2]
+
+        for k in range(10):
+
+            RTOL = 1e-2
+            ATOL = 1e-2
+
+            # An arbitrary state for the random effects
+            ref = np.random.normal(size=(model.n_groups, model.k_re)).ravel()
+
+            # Check the gradient
+            fp1 = approx_fprime(ref.ravel(), fun1, centered=True)
+            fp2 = grad(ref.ravel())
+            np.testing.assert_allclose(fp1, fp2, atol=ATOL, rtol=RTOL)
+
+            # Check the Hessian
+            he1 = approx_hess(ref.ravel(), fun1)
+            he1ld = np.linalg.slogdet(he1)[1]
+            he2 = approx_fprime(ref.ravel(), grad, centered=True)
+            he2ld = np.linalg.slogdet(he2)[1]
+            he3 = hess(ref.ravel())
```

Comment:
you could combine this with `grad` to calculate fp2 and he3 at the same time (small speedup in avoiding duplicate calculation)
Also, `rev` is already `ravel`ed AFAICS